task: initial commit adding DESIGN.md and job.proto (JobWorker service)

bucknercd opened the pull request:
- DESIGN.md – detailed service architecture, goals, and technical design considerations.

- job.proto – gRPC service definition for job lifecycle management, including start, stop, getStatus, and output streaming endpoints.

bucknercd committed:
task: initial commit adding DESIGN.md and job.proto (JobWorker service)

rosstimothy commented on DESIGN.md:
Let's simplify this to always tail from the beginning of the output, no need to make clients aware of offsets or supporting multiple modes.

bucknercd replied:
will do

rosstimothy commented on DESIGN.md:
```diff
@@ -0,0 +1,633 @@
+
+- Start/stop/query status of arbitrary processes.
+- Stream process stdout/stderr. Data will persist across server reboots.
+- Semi-persistent storage of job stream info. A TTL (time to live) for exited process will be established
```
Let's omit the TTL and keep all job information for the life of the server.

bucknercd replied:
will do

rosstimothy commented on DESIGN.md:
One of the requirements is that _all_ output is consumed. By enforcing these limits I don't think that requirement can be satisfied.

bucknercd replied:
will do

rosstimothy commented on DESIGN.md:
> then starts a job with cgroup limits, or without

One of the requirements of the challenge is that _all_ jobs have limits enforced.

bucknercd replied:
will do

rosstimothy commented on DESIGN.md:
Again, please drop the TTL and keep job resources for the entire duration of the process.

bucknercd replied:
will do

rosstimothy commented on DESIGN.md:
Again, please drop the offset and support for tail or oneshot mode.

bucknercd replied:
will do

rosstimothy commented on DESIGN.md:
For simplicity, feel free to hardcode the limits instead of consuming them from an external config source.

bucknercd replied:
Will hardcode if a config becomes unwieldy. Also, since I will use cgroups  for all jobs , I will let cgroups take care of limiting. So many of these setting will go away.

rosstimothy commented on DESIGN.md:
Where does this requirement come from? Why 15%? Is this something that could be omitted for simplicity?

bucknercd replied:
will fix

rosstimothy commented on DESIGN.md:
What happens if the entire file is consumed and the process is still running and might write more output to the file?

bucknercd replied:
I will add to the doc that when a job is terminated, a literal  `EOF` will be written to the job itself indicating that there is no more output. When the client reads, it will try to read in chunks and if there is no `EOF` on said logfile then it will wait for more streaming data until it hits a timeout (clientside). I will indicate this in the document.

bucknercd replied:
Update: I will drop the clientside timeout and drop the use of literal `EOF`. I will just rely on closing the file descriptor. When the file descriptor has been closed, the client will receive an end of stream notification.

rosstimothy replied:
I wasn't so much asking how you would handle things when a process terminated. I was trying to figure out how you would handle discovering new output. Per the challenge requirements:

> Discovering new output should be efficient, avoid busy-waiting or polling.

If the entire content of the file is consumed, but the process is still running, how will you efficiently discover new output?

bucknercd replied:
Ah yes. I updated the doc, but not here. It  will, after reading to EOF, block on `inotify` and resume reading; no busy polling.

rosstimothy commented on DESIGN.md:
What guarantees do we have that the child didn't change it's process group? Is there a more reliable way to terminate child processes?

bucknercd replied:
Since I will now use cgroups on ALL processes I will use `cgroups.kill` and before launching a process I will set pdeath signal so that the child process will get a SIGKILL if the parent dies.

rosstimothy commented on DESIGN.md:
Does this imply that the process is added to the cgroup _after_ it has started? If so, doesn't that provide a window of time for a malicious process to evade limits altogether?

bucknercd replied:
I will start a process with ptrace option so that its not running, move in cgroup, and then detach ptrace and let it run

rosstimothy commented on DESIGN.md:
Are pipes needed to write the output to files? Can you think of a way to do this without copying in goroutines?

bucknercd replied:
I don't need pipes, if this was C I'd just use `dup2` to make stdout and stderr of a process go directly to files. There is a way to do this in go. I'd just open these files up in APPEND mode, get a file descriptor and then just set cmd.stdout and cmd.stderr to point to these.   Ill update the comment.

rosstimothy commented on DESIGN.md:
As mentioned above, the order of events here is problematic.

bucknercd replied:
Will fix

bucknercd committed:
feat(design): enforce per-job cgroups with ptrace safety, simplify streaming

- Remove offsets/tail + TTL; streams always start at beginning
- Apply cgroups to all jobs; join cgroup at exec-stop (ptrace to avoid any time a proc has no limits)
- Use Setpgid + Pdeathsig; write stdout/stderr directly from process to file (no pipes)
- Stop sequence: SIGTERM to PGID, then cgroup.kill if needed
- Hardcode simple default limits (CPU 500m, Mem 100M, IO low)
- Update CLI examples and wording to match behavior
- Move proto to proto/job.proto; drop offsets; set go_package
- Add scripts/generate_proto.sh; delete old job.proto

Addresses PR #1 feedback.

rosstimothy commented on proto/job.proto:
```diff
@@ -0,0 +1,99 @@
+
+message StreamOutputResponse {
+  bytes  chunk                   = 1; // up to server chunk size (e.g., 32KB)
+  bool   end_of_stream           = 2; // true on final message before stream completes
+  uint32 server_chunk_size_bytes = 3; // informative; constant across the stream
```
I don't know that these two items add much value. There is a built in mechanism in gRPC to indicate when a stream has closed. By and large the clients shouldn't need to know about chunk sizing and have no recourse to do anything with the information.

bucknercd replied:
will fix

rosstimothy commented on DESIGN.md:
Storing metadata on disk is totally fine, but, did you consider keeping this information in memory? Would doing so simplify your implementation any?

bucknercd replied:
This would simplify the design, yes as I wouldn't have to write any of this to disk and also maintain synced with whats in memory. I also would not have to worry about any case of trying to have any metadata persistence as the only thing that would persist would be the stdout and stderr. 

The con is a server restart would wipe all job metadata which is perfectly fine for this evaluation. I will update the DESIGN.md accordingly.

rosstimothy commented on DESIGN.md:
What benefit does the channel provide if this is already running in a dedicated goroutine?

bucknercd replied:
Good point. The extra per-client channel doesn’t add value here since each gRPC StreamOutput already runs in its own goroutine. I’ll remove the intermediate channel and have the handler read the log file and Send() directly on the stream. We’ll rely on gRPC flow control for backpressure and stream.Context() for cancellation (ie. client Ctrl-C).

rosstimothy commented on DESIGN.md:
Feel free to omit graceful termination to simplify things.

bucknercd replied:
Ok. Understood. I'm switching `StreamOutput` to an event‑driven tail: no polling. The handler reads to EOF, then blocks on `inotify` for the log file. On events it resumes reading and sends chunks directly on the gRPC stream.

rosstimothy replied:
Was this reply intended for another thread? I was referring to dropping SIGTERM and relying solely on cgroup.kill here to cut scope.

bucknercd replied:
Yes. Another thread for how to discover new data without polling.

bucknercd replied:
I updated the section to go straight to `cgroup.kill` and omitted the `SIGTERM` part

rosstimothy commented on DESIGN.md:
I'm not certain that this mechanism will achieve the desired goal of avoid cgroup races. Are there any other ways that you can think of the achieve the desired behavior?

bucknercd replied:
Ok in order to avoid cgroup races, I could use `syscalls` to `fork` + `exec` and swiftly move the child `pid` into its proper cgroup while the kernel is still handling the `exec` syscall not **after** userland execution has already begun.  This should close that window between `exec` returning and placing `pid` in cgroup.

rosstimothy replied:
Is there any way to launch the child process directly in the cgroup? That seems like the most secure way to avoid any opportunity to escape the cgroup and enforced limits.

bucknercd replied:
I can do`SYS_CLONE` using  `CLONE_INTO_CGROUP` but it will be lower level as `SysprocAttrs` doesn't support this behavior yet.

bucknercd replied:
So I will rewrite the logic as:
- setup stdout/stderr with dup2
-  setup cgroup clone args
- use syscall to call the `SYS_CLONE3` and `CLONE_INTO_CGROUP` flag
- the child process will be placed in a chroot jail
- its permissions dropped to nobody,nogroup
-  set the death signal via `prctl`
- reap the process

**Note** 
I can't use a higher level construct as I need access to the `SYS_CLONE3`, `CLONE_INTO_CGROUP` syscall, arg.
Please let me know if this is ok

rosstimothy replied:
Are you sure that Go and `SysprocAttrs` don't support this functionality?

bucknercd replied:
Oh. I think it does actually. Not full clone args support but what we need to do this yes. This will simplify the code quite a bit so its not so low level. 

I will implement this same flow but with Go's higher constructs.
@rosstimothy

rosstimothy commented on DESIGN.md:
Overall you've done a good job coming up with a mechanism to identify users and determine what they have access to. I question whether the command lists is a more complicated authorization scheme than is needed for this challenge. It will require parsing the provided commands and arguments, and could be worked around by a clever user. 

Is there any other scheme you can think of that is both simpler, and still allows for admins to perform some actions and users to perform other actions?

russjones replied:
I second this. If you really want to implement this scheme, you can do it.

However both @rosstimothy and I want to save you some time. Feel free to implement a much simpler RBAC scheme.

bucknercd replied:
Yes.  I will remove the allowlist entirely and just have a denylist for all users. Only difference is how they are run. `nobody` for normal user and `1000:1000` for `admin` users.

russjones commented on DESIGN.md:
1. How are you generating these 5 bytes of random (what is the source)?
2. Isn't 5 bytes too small? Seems like this is something an attacker could brute force.

Can you provide your reasoning for this scheme to generate IDs?

bucknercd replied:
For the purpose of this challenge I wanted to have the `jobid` be short for the end user. Ideally I wanted to use something like `UUIDv7`.

1. Note: Im using the `crypto/rand` package which automatically seeds the rand function.
```go
import (
	"crypto/rand"
	"math/big"
	"strings"
)

const letters = "abcdefghijklmnopqrstuvwxyz0123456789"

func randomString(n int) string {
	var sb strings.Builder
	for i := 0; i < n; i++ {
		num, _ := rand.Int(rand.Reader, big.NewInt(int64(len(letters))))
		sb.WriteByte(letters[num.Int64()])
	}
	return sb.String()
}
```
2. I believe 5 bytes of random is too small yes. I could extend this to 10 bytes of random to make it more robust.

bucknercd replied:
I will make it as follows:
- 6 bytes to indicate part of timestamp so its sortable
- 10 bytes randomness

rosstimothy replied:
You may use a UUID package if you wish instead of hand rolling your identifier.

bucknercd replied:
Ok will use UUIDv7

bucknercd commented:
Understood. I will remove the allowlist entirely and just have a denylist for all users. Only difference is how they are run. nobody for normal user and 1000:1000 for admin users.

bucknercd committed:
feat(proto)!: simplify StreamOutputResponse; adopt ForkExec, inotify, and denylist in design

- Remove StreamOutputResponse.end_of_stream and server_chunk_size_bytes; rely on gRPC stream close and server-controlled chunking
- Switch process launch design to syscall.ForkExec; open logs first, set ProcAttr.Env to "PATH="+safePATH, Setpgid/Pdeathsig/Credential
- Create cgroup and apply limits before forking; immediately write child PID to <cg>/cgroup.procs (no ptrace/SIGSTOP)
- Streaming: direct file→gRPC; no per-client channels; EOF→block on inotify (no busy polling)
- Security: unify to a single denylist for all roles; run as nobody (user) vs 1000:1000 (admin); shells on denylist
- Metadata: keep in memory only (logs persist on disk)
- Job IDs: 16 chars = 6-char base36 timestamp (fixed width) + 10 crypto-random [a-z0-9]
- Config: add safePATH (ie. /usr/bin:/bin) and port; small wording/typo fixes

zmb3 commented on proto/job.proto:
```diff
@@ -0,0 +1,97 @@
+}
+
+message StopJobResponse {
+  string      job_id   = 1;
```
Probably no need to return the ID here since the client already specified it in the request.

bucknercd replied:
will fix

zmb3 commented on proto/job.proto:
```diff
@@ -0,0 +1,97 @@
+
+message StreamOutputRequest {
+  string job_id        = 1;
+  bool   stream_stderr = 2; // false (default) = stdout, true = stderr
```
Does `true` mean:
a) stdout _and_ stderr, or
b) stderr only

Depending on your answer, maybe either renaming the field or using an enum instead of a bool could have made it more clear.

bucknercd replied:
Noted. For better clarity I will implement an Enum for this with `STDOUT` and `STDERR`. 
The `streamOutput` call will stream `STDOUT` if nothing is specified. If `--stderr` is specified then `streamOutput` will stream `STDERR`. This is for simplicity. **Only** `STDOUT` or `STDERR`, not both.

zmb3 commented on DESIGN.md:
````diff
@@ -0,0 +1,471 @@
+  - I/O: low
+```bash
+# Start a job
+jobctl start --cmd "ls / -lah"
````
By putting the command and its arguments into a single string like this, your program becomes responsible for parsing the string into a command and a set of arguments.

This is not necessarily bad, but feels like unnecessary complexity. We're basically asking the user to go out of their way to make sure the shell _doesn't_ do this parsing (by quoting the whole string) and then doing it ourselves.

Any opportunity to simplify?

bucknercd replied:
I can have the user invoke like so: `jobctl start ls / -lah`

- No more wrapping everything in quotes and no more `--cmd`
-  Makes user's experience nicer.
- The command + args will still be treated as one string. Its just that index 0 will be `command` and the rest will be args. Not very much parsing. Very straightforward. 
- All the limit flags will be optional and explicit (ie. --cpu=500m --memory=1G --io=med)
-  This can be accomplished via Cobra

This should simplify. I will implement. Please let me know if that is ok.

zmb3 replied:
Sounds good

zmb3 commented on DESIGN.md:
```diff
@@ -0,0 +1,471 @@
+  --io=low
+
+# Stream the output (stdout)
+jobctl stream --id=xxxxxxxx
```
If `--id` is always required, do we even need a flag for it? (Same question applies to `--cmd` in the `start` subcommand.)

bucknercd replied:
Will make `--id` required and thus no need to have a flag for it.

zmb3 commented on DESIGN.md:
```diff
@@ -0,0 +1,471 @@
+func randomString(n int) string {
+    var sb strings.Builder
+    for i := 0; i < n; i++ {
+        num, _ := rand.Int(rand.Reader, big.NewInt(int64(len(letters))))
```
Do we need big integers here? Why not `rand.Intn`?

bucknercd replied:
will use UUIDv7

zmb3 commented on DESIGN.md:
```diff
@@ -0,0 +1,471 @@
+| Function             | Description                                             |
+| -------------------- | ------------------------------------------------------- |
+| `Start()`            | Starts a process via setpgid(), takes in an executable and args. Sets job status, cgroup limits and executes the process. See **Start** for details.  It also sets up the process' stdout and stderr and to be read from and written to files on disk. These file names should map up to the job id assigned and also on disk as well. Then wait for job to finish, or be terminated,  ...
+| `Stop()`             | Sends `SIGTERM` to the process group for as an initial try. If the process (or children) is still running after the timeout, it writes `1` to `/sys/fs/cgroup/jobs/<jobid>/cgroup.kill` to forcefully terminate it and remove the cgroup.
```
Feel free to skip straight to `cgroup.kill` in order to simplify the implementation.

bucknercd replied:
will remove the sending of the `SIGTERM` signal.

zmb3 commented on DESIGN.md:
```diff
@@ -0,0 +1,471 @@
+- The `Start()` function will take in a command and its arguments, along with resource limits. It will then:
+  - Generate a unique job ID
+  - Create a cgroup for the job and apply resource limits
+  - Create a new process using `syscall.ForkExec()` with the following attributes:
```
This seems lower-level than necessary. The syscall package docs even say:

>  The primary use of syscall is inside other packages that provide a more portable interface to the system, such as "os", "time" and "net". Use those packages rather than this one if you can. 


Can we use `os/exec` instead?

bucknercd replied:
Will use `os/exec.Cmd` instead so its more portable and a init binary for launching the actual process.

bucknercd replied:
~~As stated above, I can't put a process directly into a `cgroup` unless I use a `SYS_CLONE3` system call which I can't do via `os/exec` at the moment.  If the `cgroup` race condition window wasn't an issue then I could. I researched some and even the go implementations like `containerd`  have this same issue. There is a moment between `fork` and `exec` where the `pid` is being moved into the `cgroup` that is an attack window. The Linux kernel 5.7+  the `clone` syscall now how `CLONE_INTO_CGROUP` as to resolve this issue.  The will remove the window completely but complicates the code and makes it low level.~~

Update: I will use os/exec.Cmd with Go1.21+ . Doc will reflect details

zmb3 commented on DESIGN.md:
```diff
@@ -0,0 +1,471 @@
+        - stdin: `/dev/null`
+        - stdout: `<jobdir>/stdout.log`
+        - stderr: `<jobdir>/stderr.log`
+  - Immediately move the process to the cgroup by writing its PID to `/sys/fs/cgroup/jobs/<jobid>/cgroup.procs`
```
This means there will be a non-zero amount of time where the process is running but is not yet in the cgroup.

Any concerns about that approach?

bucknercd replied:
Ok. Understood. By setting up the process for running using low level fork + exec , I had lowered the window of opportunity for cgroup races. (Plus not as portable) I see the goal now. Not a very slim window; **no** window. 

I’m switching over to `os/exec.Cmd` instead of `syscall.ForkExec()`.
We still get all the knobs we care about in SysProcAttr (Setpgid, Pdeathsig, Credential, etc.).

For the cgroup timing issue, I’m gonna run everything through a tiny init helper binary.
That init process will:
- Drop itself into the right cgroup before doing anything else.
- `exec` the real target, so the user’s process starts inside the cgroup from the very first instruction.
- This keeps our control over the process, closes the `process is running before it’s in the cgroup` window

I will update DESIGN.md

bucknercd replied:
~~As stated above, I can't put a process directly into a `cgroup` unless I use a `SYS_CLONE3` system call which I can't do via `os/exec` at the moment.  If the `cgroup` race condition window wasn't an issue then I could. I researched some and even the go implementations like `containerd`  have this same issue. There is a moment between `fork` and `exec` where the `pid` is being moved into the `cgroup` that is an attack window. The Linux kernel 5.7+  the `clone` syscall now how `CLONE_INTO_CGROUP` as to resolve this issue.  The will remove the window completely but complicates the code and makes it low level.~~

Update: I will use os/exec.Cmd with Go1.21+ . Doc will reflect details

@zmb3

zmb3 commented on DESIGN.md:
```diff
@@ -0,0 +1,471 @@
+  - Create `/sys/fs/cgroup/jobs` if it doesnt exist yet
+2. Inside  `<jobid>` directory:
+  - set `memory.max` in bytes (one token)
+  - set `cpu.max` to "<quota> <period>" where they are in microseconds. (2 tokens)
```
How are you going to get `<quota>` and `<period>` if the user provided a single string like `500m`?

bucknercd replied:
If `--cpu` is set to `500m` then in cgroups v2 then it would be set as so `cpu.max = 50000 100000`  for `quota` and `period` respectively. I am going to hardcode `period` and adjust `quota` based on the CLI `cpu` flag.

I will add a mapping to DESIGN.md

zmb3 commented on DESIGN.md:
```diff
@@ -0,0 +1,471 @@
+  - set `memory.max` in bytes (one token)
+  - set `cpu.max` to "<quota> <period>" where they are in microseconds. (2 tokens)
+    - cpu.max is written as "<quota> <period>" in microseconds. Typical period is 100000µs (100ms), and quota ≤ period.
+  - set `io.max` to "<device> <rbps> <wbps>" (3 tokens). An example: `8:0 rbps=1048576 wbps=1048576`
```
Similarly, how will the low/med/high class from your proto map to `rbps` and `wbps`?

bucknercd replied:
`low`, `med` and `high` will be mapped to `1M`, `10M`, `0`(unlimited) respectively for both `rpbs` and `wbps`. I will determine the device which is on the root partition and use that in `io.max`.

I will add a mapping to DESIGN.md.

zmb3 commented on DESIGN.md:
```diff
@@ -0,0 +1,471 @@
+    - Note: The major(8) and minor(0) will try to be determined based on device where the root(/) directory is mounted on.
+3. Immediately after `ForkExec` returns (in the parent), write the child PID to `/sys/fs/cgroup/jobs/<jobid>/cgroup.procs`.
+4. When the process exits, delete the cgroup directory `/sys/fs/cgroup/jobs/<jobid>`.
+5. If the process is still running, write `1` to `/sys/fs/cgroup/jobs/<jobid>/cgroup.kill` to forcefully kill the process and remove the cgroup.
```
Maybe I'm just confused by the order in which this is written, but if you've deleted the cgroup directory in step 4 there won't be anything to write here.

bucknercd replied:
This is a logical error in the documentation. I will correct since yes it doesn't make sense this way since the cgroup directory would be deleted.

I will fix.

zmb3 commented on DESIGN.md:
```diff
@@ -0,0 +1,471 @@
+  - `GetStatus`
+  - `StreamOutput` (streaming gRPC)
+- Secures:
+  - mTLS setup using TLS 1.3
```
How will the server know which client certificates to trust?

Similarly, how will the CLI client know which client certificate to use?

bucknercd replied:
I will have a script to autogenerate certs for clients based on their own input. They all chain up to one `CA`. If the client cert chains up to the `CA`, then its trusted by the server.

The CLI client will use the client cert in its own generated directory with its private key as well. So a mechanism for starting up a client must be implemented. For testing purposes could be hardcoded but a tiny client config could help as well that could indicate the username.

zmb3 commented on DESIGN.md:
```diff
@@ -0,0 +1,471 @@
+### 5. AuthN/Z
+- The client must authenticate the server **and** the server must authenticate the client. This means that the client **and** the server must trust each other via CA's. For simplicity, the server will have the client's CA's when running.
+- `CN` will be used for `<username>`
+- `O` will be used for `<role>-<group>`
```
Up above you said

> authorization done via CA. O=admin|user

This seems to contradict with the `<role-group>` notation mentioned here.

bucknercd replied:
Will fix. Going with CN=username. Will no longer use role.

zmb3 commented on DESIGN.md:
```diff
@@ -0,0 +1,471 @@
+  - TLS 1.3 only
+  - Strong cipher suite (TLS\_AES\_256\_GCM\_SHA384 preferred)
+- **Authorization policy**
+  - In the client certificate itself, `CN` will be the `username` and `O` will be the `role-group`
```
After reading through this a couple times I'm still not totally clear on the authorization model. Maybe having both roles _and_ user groups is more complicated than what we need here? I'd welcome any opportunity to simplify.

bucknercd replied:
Yes. I will simplify the AuthZ scheme so that `CN` will be for username. Each individual `user` will **only** be able to interact with their jobs. All processes will run in chroot jail as `nobody` and I will get rid of the denylist entirely.

zmb3 commented on DESIGN.md:
```diff
@@ -0,0 +1,471 @@
+
+#### NOTE: All lists will be hardcoded in the implementation
+
+### Denylist for `admin` users and normal users
```
I'd consider alternate approaches. Allow/deny lists are really easy to circumvent - there's just too many corner cases.

For example: `ln /bin/bash /home/me/myunrestrictedshell` or `apt install some-shell-thats-not-on-thelist` and then the deny list is completely ineffective from there.

bucknercd replied:
Ok. An alternative approach would be to use process isolation via chroot for launching the process.  This way the attack surface will be reduced greatly. On a similar note, I will drop permissions to nobody for all users and simplify the RBAC scheme.

zmb3 commented on DESIGN.md:
```diff
@@ -0,0 +1,471 @@
+5. If the process is still running, write `1` to `/sys/fs/cgroup/jobs/<jobid>/cgroup.kill` to forcefully kill the process and remove the cgroup.
+
+
+### 3. `server` (gRPC API)
```
I'd love to see what the Go API looks like at the library level, too.

bucknercd replied:
Understood. I will a simple example of how to use the joblib directly (not via the manager package which would wrap around it)

Something like:

```
import manager

m := manager.NewJobManager(...)
jobID, err := m.StartJob([]string{"/bin/sleep", "5"}, joblib.ResourceLimits{CPU: "500m"})
status, _ := m.GetStatus(jobID)
stream, _ := m.StreamOutput(jobID, false)
m.StopJob(jobID)
```

@zmb3

zmb3 commented on DESIGN.md:
```diff
@@ -0,0 +1,471 @@
+5. If the process is still running, write `1` to `/sys/fs/cgroup/jobs/<jobid>/cgroup.kill` to forcefully kill the process and remove the cgroup.
+
+
+### 3. `server` (gRPC API)
```
This section doesn't provide much value IMO. You've already discussed mTLS in other parts of the document, the gRPC API is covered in the protobuf spec, and authn/z has its own dedicated section too.

bucknercd replied:
Will remove.

zmb3 commented on DESIGN.md:
```diff
@@ -0,0 +1,471 @@
+
+## CLI Examples
+### Streaming Note
+- All the `stream` commands have `tail` like functionality. This means that if the job is still running, the stream will continue to follow the output in real-time. If the job has exited, the stdout and stderr file descriptors will be closed and the stream will end.
```
```suggestion
- The `stream` command has `tail` like functionality. This means that if the job is still running, the stream will continue to follow the output in real-time. If the job has exited, the stdout and stderr file descriptors will be closed and the stream will end.
```

bucknercd replied:
Will fix

zmb3 reviewed (commented):
There's a lot of good content here but I think it could be organized  or communicated better. I had to read through it a few times before I felt like I picked everything up.

bucknercd committed:
docs: align DESIGN.md and job.proto with updated architecture

- DESIGN.md:
  - Clarify simple authorization uses CN from client CA as username.
  - Replace denylist approach with chroot jail for restricted commands.
  - Simplify AuthZ policy, remove role/group execution rules.
  - Update CLI examples to use positional arguments instead of flags.
  - Add chrootRoot to config defaults.
  - Expand Start() section to detail init binary responsibilities.
  - Enhance cgroups section with CPU/memory/IO mapping table and examples.
  - Add mTLS certificate trust & selection summary table.
  - Remove outdated admin/normal user denylist section.

- proto/job.proto:
  - Add descriptive comments for enums/fields to match DESIGN.md.
  - Remove unused role/group fields from JobMetadata.
  - Clarify ResourceLimits formats and mapping to cgroups.
  - Update StartJobResponse job_id format to 16-char sortable string.
  - Switch StreamOutputRequest from bool to StreamTarget enum.
  - Document streaming replay and multi-client support.

strideynet commented on proto/job.proto:
```diff
@@ -0,0 +1,122 @@
+
+// Output target to stream.
+// Defaults to STDOUT if not set.
+enum StreamTarget {
```
Nit: The Protobuf best practices advises against the zero value of the enum being semantically meaningful and suggest the inclusion of an explicit `UNSPECIFIED` value - https://protobuf.dev/best-practices/dos-donts/

bucknercd replied:
Will add the unspecified for 0

rosstimothy commented on DESIGN.md:
````diff
@@ -0,0 +1,511 @@
+```go
+//Supervisor (parent process)
+//Just starts the wrapper and handles logging; no sandboxing here because the wrapper will do it.
+cmd := exec.Command("/usr/local/bin/init-wrapper", append([]string{jobId, targetCmd}, args...)...)
````
I think that the use of `/usr/local/bin/init-wrapper` would be in violation of the challenge requirements. Per the challenge spec:

> The server should also not rely on any shell scripts, external binaries or use containers to execute jobs.

bucknercd replied:
~~Understood. I finally found out how to have **zero** window. I would have to use a `clone()` syscall with `CLONE_INTO_CGROUP`.   It will require kernel version 5.7+  though.~~

~~Summary
As stated above, I can't put a process directly into a cgroup unless I use a `clone3()` syscall which I can't do via `os/exec` at the moment. If the `cgroup` race condition window wasn't an issue then I could. I researched some and even the go implementations like `containerd` have this same issue. They support kernel versions < 5.7. There is a moment between fork and exec where the `pid` is being moved into the `cgroup` that is an attack window. In the Linux kernel 5.7+ the `clone3` syscall using `CLONE_INTO_CGROUP` now resolves this issue. That will remove the window completely but complicates the code and makes it low level.~~

 Update: I will use os/exec.Cmd with Go1.21+ . Doc will reflect details

@rosstimothy

bucknercd committed:
doc(design): clarify joblib start process, status flow, IO device source; update proto enums and streaming notes

- DESIGN.md:
  - Clarify job status flow and definitions (Exited, Stopped, Failed).
  - Replace custom job ID generation with UUIDv7.
  - Document Go 1.21+ and Linux 5.7+ requirements for UseCgroupFD/CLONE_INTO_CGROUP.
  - Expand `Start()` description with SysProcAttr, chroot jail, privilege drop, and logging.
  - Change IO device reference from `/` to `/opt` consistently.
  - Remove obsolete init binary workflow.
  - Note that PID writing to cgroup.procs is unnecessary with UseCgroupFD.

- proto/job.proto:
  - Add `STREAM_TARGET_UNSPECIFIED` to StreamTarget enum.
  - Add comments clarifying chunk semantics in StreamOutputResponse.

strideynet commented on proto/job.proto:
```diff
@@ -0,0 +1,125 @@
+// Defaults to STDOUT if not set.
+enum StreamTarget {
+    STREAM_TARGET_UNSPECIFIED = 0; // No target explicitly set
+    STDOUT = 1;
```
Nit: Should STDOUT and STDERR be prefixed with the enum name as well as you have done in JobStatus?

bucknercd replied:
Will fix

rosstimothy commented on DESIGN.md:
```diff
@@ -0,0 +1,541 @@
+| `Start()`            | Creates a cgroup via `cgroup` library, sets up the command with resource limits, and starts the process in a chroot jail. It will also handle logging to stdout/stderr files. The function will return a job ID.
+| `Stop()`             | Has cgroup library write `1` to `/sys/fs/cgroup/jobs/<jobid>/cgroup.kill` to forcefully terminate and then remove the cgroup.
+| `Status()`           | Return the job status response 
+| `GetStatusResponse()`| Returns a structured status response with job ID, current state (running, exited, etc.), and exit code (if available). Used by both CLI and gRPC server.
```
Are both Status and GetStatusResponse needed? What is the difference between the two?

bucknercd replied:
One is pure status for the manager to use and the other is just **all** response statuses in one. The pure `status()` is redundant. I will fix this when implementing. I will delete this from design doc

rosstimothy commented on DESIGN.md:
```diff
@@ -0,0 +1,541 @@
+
+2. **Build the Command**
+    - Open log files for stdout and stderr
+    - Build the command using `exec.Command` with the provided command and arguments using a safe `PATH` environment variable
```
No need to worry about PATH or other environment variables. Feel free to omit to simplify and reduce scope.

bucknercd replied:
Ok. thats good to know. I wont worry about using a safe `PATH`

rosstimothy commented on DESIGN.md:
```diff
@@ -0,0 +1,541 @@
+
+3. **Use SysProcAttr to set up the child process**
+  - Set `UseCgroupFD` to true and pass the cgroup file descriptor
+  - Set `Chroot` to the chroot jail directory (e.g., `/opt/jobroot`)
```
No need to worry about chroot. Feel free to omit to simplify and reduce scope.

bucknercd replied:
So I can just run as `nobody:nogroup` any command?

From the feedback I got about the denylist/allowlist, I had come up with this. 
I was thinking about implementing a chroot jail with limited binaries + deps.

rosstimothy replied:
No need to support chroot jail. For this exercise it is fine to run the command passed in verbatim without any extra handling or security considerations. All we are looking to see if that you can identify some of the risks that poses and document how to mitigate them in your design.

bucknercd replied:
OK I went ahead and just removed all references to it from the doc.

rosstimothy commented on DESIGN.md:
````diff
@@ -0,0 +1,541 @@
+| **Server** | - Loads **trusted client CA certificates** from a fixed location on disk.(`ca.crt`)<br>- Configured with `ClientAuth: RequireAndVerifyClientCert`.<br>- Verifies presented client cert chains to a trusted CA and is valid (expiry, key usage). | *Not applicable* — server only verifies client certs. |
+| **CLI Client** | - Loads **server CA certificate** from fixed location on disk.(`ca.crt`)<br>- Uses this to verify the server’s certificate during TLS handshake. | - Loads **its own client certificate** (`.../bob/client.crt`) and private key (`.../bob/client.key`) from fixed, hardcoded paths.<br>- The CN in the cert acts as the `username` for AuthZ.<br>- Always uses the same keypair for all req ...
+
+**Notes:**
+- All trust material is **static**
+- `CN` in the client certificate = `username`
+- Trust is based solely on local CA files.
+##### NOTE: All lists will be hardcoded in the implementation
+---
+### Limited list of binaries for chroot jail
+#### Will allow more in the future
+```bash
+/bin/ls
+/bin/echo
+/bin/cat
+/bin/sleep
+/bin/date
+...
+...
+...
+```
+
+- Will need to include the dependencies of these binaries as well
````
Does the allow list here still apply? Is this left over from the previous authorization scheme?

bucknercd replied:
Since I wanted to use a chroot jail, I will need a list of allowed binaries. This is not an allowlist like before but what binaries will actually live in the chroot jail where the commands will be run. I went this route since I got feedback about allowlist/denylists being easily bypassed.

rosstimothy replied:
As mentioned before I think chroot jail can be a future consideration and is not a requirement of this challenge.

bucknercd replied:
Ok. I will go with no chroot jail and just run everything as `nobody:nogroup`

rosstimothy commented on DESIGN.md:
```diff
@@ -0,0 +1,541 @@
+### 5. AuthN/Z
+- The client must authenticate the server **and** the server must authenticate the client. This means that the client **and** the server must trust each other via CA's. For simplicity, the server will have the client's CA's when running.
+- `CN` will be used for `<username>`
+- Process streaming or getting statuses, reading of any job will be limited to each user.
```
Can you help clarify the authorization scheme for me? 

From [above](https://github.com/bucknercd/jobworker/pull/1/files#diff-3dc5dd454e080eb849ee5efaf79df2585fbe0a06804d69249b4c81da05a63875R86-R89), the authorization scheme is said to be

> A user can **only** can perform operations such as stream output from any job as long as they have the job id 

Does that contradict what is being described here?

bucknercd replied:
No, I just didn't specifically address that aspect of things here. But, yes, in order to stream a `job id` will be needed.

bucknercd replied:
Will fix

bucknercd committed:
doc(design): remove hardcoded safe PATH and refine CLI access rules

- Removed references to hardcoded PATH env from DESIGN.md
- Deleted obsolete `Status()` entry in server method table
- Clarified CLI client access: job streaming/status requires same creator user
- Renamed StreamTarget enum values for clarity (`STREAM_TARGET_STDOUT`, `STREAM_TARGET_STDERR`)

bucknercd committed:
docs: remove chroot jail references and update process execution details

- Removed all mentions of chroot jail from design
- Updated process execution section to clarify direct execution without shells
- Simplified privilege dropping section to only run as nobody:nogroup
- Removed hardcoded limited binary list for chroot
- Adjusted `Start()` function description to remove chroot-related steps

bucknercd committed:
doc: clarify mTLS AuthN/Z scheme and trust boundaries

- Added explicit details on client/server CA requirements
- Documented that server authorization is based on CN in client certificate
- Defined trust boundary requiring all requests over mTLS with valid cert
- Added step-by-step Auth Flow for user, job ID, and ownership checks

bucknercd committed:
feat(core): add chroot + joblib, logging, cgroup packages and more

- Introduce `scripts/chroot.sh` with build|clean|nuke lifecycle
- Wire up dedicated `Makefile` targets (`chroot`, `chroot.clean`, `chroot.nuke`)
- Exclude chroot ops from global `make clean` (safety)
- Add initial `internal/joblib` with Job struct:
  - cgroup manager stub
  - chroot jail support (`/opt/jobroot`)
  - stdout/stderr log file handling
  - lifecycle states and transitions
- Add `internal/logging` package for file-backed logger
- Add initial `internal/cgroups` stub (Create/Delete signatures)
- Add `certs/Makefile` for CA/server/user cert generation
- Update proto generation script to use `clean` flag
- Update DESIGN.md with chroot jail + nobody:nogroup hardening
- Add go.mod/go.sum with grpc + protobuf deps
